### PR TITLE
Jwoo/ios visibility update

### LINKED ISSUE #6149: [Authoring] [iOS]When width set to "auto" there is a misalignment if its previous item is hidden
# Target Platform

Please list the platform(s) that your application is targeting and remove platforms you are not targeting:
- iOS

# Target Application

Please provide the name of the application that is rendering the card (ex: Outlook, Teams, etc)
Teams

# Problem Description

Please enter a description of the issue. If you just have a question, please post [on StackOverflow instead](https://stackoverflow.com/questions/tagged/adaptive-cards)
when there are more than 1 item in the "columns" and each item has their width set to "auto", the items don't move to the left when the previous item in the queue gets hidden. The designer does move the items to the left(check the json and the repro steps). Note below in the screenshot, iPhone_water shows the image is somewhere in the middle, when you tap on "water" vs the designer_water, when you tap on "water" has the image left aligned.

# Screenshot

Please provide a screenshot of the problem (if possible)
Designer_water
iPhone_water

# Card JSON

Please provide the JSON of the Card that is being rendered
{"body":[{"type":"TextBlock","text":"Press the buttons to toggle the images!","wrap":true},{"id":"textToToggle","isVisible":false,"type":"TextBlock","text":"Here are some images:"},{"columns":[{"width":"auto","items":[{"height":"auto","style":"person","id":"imageToToggle","size":"medium","isVisible":false,"type":"Image","url":"https://picsum.photos/100/100?image=112"}],"type":"Column"},{"width":"auto","items":[{"size":"medium","id":"imageToToggle2","isVisible":false,"url":"https://picsum.photos/100/100?image=123","height":"auto","type":"Image"}],"type":"Column"}],"type":"ColumnSet"}],"version":"1.2","type":"AdaptiveCard","actions":[{"title":"Toggle!","targetElements":["textToToggle","imageToToggle","imageToToggle2"],"type":"Action.ToggleVisibility"},{"title":"Also Toggle!","targetElements":["textToToggle","imageToToggle","imageToToggle2"],"type":"Action.ToggleVisibility"},{"title":"Show!","targetElements":[{"isVisible":true,"elementId":"textToToggle"},{"isVisible":true,"elementId":"imageToToggle"},{"isVisible":true,"elementId":"imageToToggle2"}],"type":"Action.ToggleVisibility"},{"title":"Hide!","targetElements":[{"isVisible":false,"elementId":"textToToggle"},{"isVisible":false,"elementId":"imageToToggle"},{"isVisible":false,"elementId":"imageToToggle2"}],"type":"Action.ToggleVisibility"},{"title":"Grain!","targetElements":[{"isVisible":true,"elementId":"textToToggle"},{"isVisible":true,"elementId":"imageToToggle"},{"isVisible":false,"elementId":"imageToToggle2"}],"type":"Action.ToggleVisibility"},{"title":"Water!","targetElements":[{"isVisible":true,"elementId":"textToToggle"},{"isVisible":false,"elementId":"imageToToggle"},{"isVisible":true,"elementId":"imageToToggle2"}],"type":"Action.ToggleVisibility"}],"$schema":"http:\/\/adaptivecards.io\/schemas\/adaptive-card.json"}

# Sample Code

Please provide any relevant code samples to help reproduce the issue

# Version of SDK

What version are you using? Ex: NuGet 1.0.2, or latest main, etc...
pod file 2.8.0-beata1

### PR DESCRIPTION
**Pull request by jwoo-msft**
# Related Issue

Fixed #6149 

# Description

* If Column width is `auto` and all of its items are hidden, column width should become zero, if `stretch`, the column can take up the space.

To implement this behavior, added a visibility manager to make the code testable and isolate the feature. ColumnView owns this manager. The column's subviews are mapped to the column view, and the mapping is registered to the rendering context.  When the toggle visibility event handler asks for the visibility changes, it checks with the context if the visibility manager facade (column view) is available, if so, it asks the facade for the visibility changes. The facade has access to all of its children and has a visibility manager which knows how to handle visibility and turns on and off the visibility.

# Sample Card
https://github.com/microsoft/AdaptiveCards/blob/main/samples/v1.2/Tests/Action.ToggleVisibilityExhaustive.json

# How Verified

How you verified the fix, including one or all of the following:
1. New unit tests are added.
2. All unit tests are run & tested with relevant json.
Old Behavior
Screen Shot 2021-08-06 at 2 46 05 PM

New Behavior when "auto"
Screen Shot 2021-08-06 at 2 44 33 PM
New Behavior when "stretch"
Screen Shot 2021-08-06 at 2 46 05 PM

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/AdaptiveCards/pull/6172)

### CODE REVIEW

**Review by paulcam206 — APPROVED** [review]

File: source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACOVisibilityManager.mm [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+    if (!viewToBeHidden.isHidden) {
+        viewToBeHidden.hidden = YES;
+    }
+
+    // get the index of the target view
+    // if the target view is not the tail,
+    // check to the right, and if the right separator and not hidden,
+    // hide the separator
+    // if the target view is tail, check to left, and if the left is separator and not hidden,
+    // nhide the separator
```
Comment (almedina-ms) [inline source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACOVisibilityManager.mm:30]: nit: typo hide

File: source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACOVisibilityManager.mm [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+        if ([prevView isKindOfClass:[ACRSeparator class]]) {
+            if (!prevView.isHidden) {
+                prevView.hidden = YES;
+            }
+        }
+    }
+
+    // only cares for tailing padding at the moment
+    // if column width is "auto", and all of its children hidden,
+    // hide column, if is "stretch", then don't hide it.
```
Comment (almedina-ms) [inline source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACOVisibilityManager.mm:52]: why is a hidden stretch column expected to occupy space? are we trying to mimic the behaviour of visibility invisible and collapsed with this behavior?

File: source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACOVisibilityManager.mm [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+    // find the index of target view and check if the view is at tail
+    // if tail, check to left for separator else check to right
+    // if trailing padding exists, the target view is tail and padding is hidden,
+    // unhide the padding regardless
+    BOOL isViewFound = NO, isLastView = YES;
+    NSInteger targetIndex = subviews.count;
+    for (NSInteger i = 0; i < subviews.count; i++) {
+        UIView *subview = subviews[i];
+        if ([subview isEqual:viewToBeUnhidden]) {
+            targetIndex = i;
```
Comment (almedina-ms) [inline source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACOVisibilityManager.mm:81]: why is this process manually done when unhiding? besides determining if the view to be shown is the last one?

File: source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACRImageRenderer.mm [hunk 4]
```diff
@@ -149,7 +149,7 @@ - (UIView *)render:(UIView<ACRIContentHoldingView> *)viewGroup
         // if we already have UIImageView and UIImage, configures the constraints and turn off the notification
         [self configUpdateForUIImageView:rootView acoElem:acoElem config:acoConfig image:view.image imageView:view];
     }
-
+
```
Comment (almedina-ms) [inline source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACRImageRenderer.mm:152]: not: extra space

File: source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACRToggleVisibilityTarget.mm [hunk 5]
```diff
@@ -38,32 +39,49 @@ - (void)doSelectAction
         NSString *hashString = [NSString stringWithCString:target->GetElementId().c_str() encoding:NSUTF8StringEncoding];
         NSUInteger tag = hashString.hash;
         UIView *view = [_rootView viewWithTag:tag];
+        UIView *separator = nil;
+        BOOL bHide = NO;
```
Comment (almedina-ms) [inline source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACRToggleVisibilityTarget.mm:43]: what does the b stand for? if the variable holds the visibility value that the element will have could it be a better idea to name the variable with something related to visible?

File: source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACOVisibilityManager.mm (hunk 2 above)
Comment (jwoo-msft) [reply source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACOVisibilityManager.mm:52]: David said that auto takes the consideration of content size. So if its children are all hidden, the column's content will become zero and doesn't occupy the space. With the stretch, the content size is not considered, and it occupies space regardless. And I think he is correct, this is in line with what we have defined for column width.

File: source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACRToggleVisibilityTarget.mm (hunk 5 above)
Comment (almedina-ms) [reply source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACRToggleVisibilityTarget.mm:43]: nevermind, the whole concept of the visibility manager uses the word 'hide' or 'hidden' so it's ok to keep it under those terms

**Review by almedina-ms — APPROVED** [review]

File: source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACOVisibilityManager.mm (hunk 2 above)
Comment (almedina-ms) [reply source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACOVisibilityManager.mm:52]: if this was defined and accepted by the team then it's ok

File: source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACOVisibilityManager.mm (hunk 3 above)
Comment (jwoo-msft) [reply source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACOVisibilityManager.mm:81]: If the view is hidden in UIStackView. it's removed from its arrangeddSubviews, and you can't call the function if you're looking for a view that's hidden.

**Pull request merged**